### PR TITLE
Convert our platform into a more AWS friendly format

### PR DESCRIPTION
**Pull request by Hiro-Nakamura**
The big goal of this revision is to generate an installation that removes any (most) of the shared files between the host computer and our running containers.

Instead of trying to use a common `config/local.js` file to share configuration data among our containers, we are using an `.env` file to generate environment variables that are passed into our containers.

These changes touch just about every repository we own to implement.

In order to test this out:
-  check out this branch of the `ab-cli` on your local machine
-  then perform a sample install

here is a sample install that already includes all the questions filled out:
```
appbuilder install env --stack=env --port=8080 --dbExpose=true --dbPort=8889 --dbPassword=rO0t --tag=master --authType=login --runtime=jh/awsReady --relayEnabled=false --pwaURL=http://pwa.site.com --siteURL=http://auth.site.com --tenant.username=admin --tenant.password=admin --tenant.email=admin@email.com --tenant.url=https://adminTenant.site.url
```

Also test it with the `--develop` tag to verify a newly installed development environment works as well.

### CODE REVIEW

**Comment by Hiro-Nakamura** [comment]
@joshAppDev Can you run some tests to verify this works and to document what changes need to be made when using `podman`?

Also, if you could document here how we will need to migrate a running site to use the new setup.

**Comment by Hiro-Nakamura** [comment]
@guyyoo Can you review the automated tests for this branch and make sure it is using the correct parameters?

specifically make sure it is including a `--db.password`  and the `--runtime=jh/awsReady` parameters.  (might need more  use the example I pasted in above as your guide)

**Review by jinzsom — APPROVED** [review]
Okay, I have fixed the test to utilize the new parameters. However, please note that the `--runtime` parameter must match the branch name of the 'ab-cli' branch. Otherwise, it will default to using `ab_runtime#master`.

**Comment by joshAppDev** [comment]
> Can you run some tests to verify this works and to document what changes need to be made when using podman?
> Also, if you could document here how we will need to migrate a running site to use the new setup.

That actually sounds like maybe an M sized sprint task. I'll bring it up during the sprint planning. As for the review, here is some feedback:

### CLI vs Git
The ideal setup in my mind is as simple as possible. Just git clone a repo, maybe copy & paste the .env settings, and then launch Docker. I really wasn't expecting ab-cli to be part of the AWS setup. Now I've never used AWS myself, but I imagine that you would type the .env settings into some Amazon interface, and it would propagate to all your EC2 instances, or something like that.

One benefit of deploying via a git repo, is that updating the site files (like docker-compose.yml) can be done by just a git pull.

### Database

Will we be using an AWS DB for our database? If we are going to scale up to multiple instances, would they need to all connect to a central DB? (Again, I've never done this myself.)

File: templates/setup/docker-compose.override.yml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    # extends:
+    #   service: common_config
+
+
+  #db: use Maria DB as our backend DB
+  <%= dbExpose ? "" : "# " %>db:
+  <%= dbExpose ? "" : "# " %>  ports:
+  <%= dbExpose ? "" : "# " %>    - "<%=dbPort%>:3306"
+  <%= dbExpose ? "" : "# " %>  # extends:
+  <%= dbExpose ? "" : "# " %>  #   service: common_config
```
Comment (joshAppDev) [inline templates/setup/docker-compose.override.yml]: Can the port numbers be read from the .env?

File: templates/setup/config/local.js [hunk 2]
```diff
@@ -39,11 +39,12 @@ module.exports = {
 
    /** Session */
    session: {
-      secret: "<%=secret%>",
+      secret: "<%=sailsSessionSecret%>",
```
Comment (joshAppDev) [inline templates/setup/config/local.js]: Can the session secret be read from the .env?

File: templates/setup/[stack]_system_monitor.js [hunk 3]
```diff
@@ -237,7 +238,7 @@ var serviceNames = {};
 // a quick lookup hash with the names of the current docker stack + service
 // name for each of our servicesToWatch.
 
-var stack = "ab";
+var stack = "<%= stack %>";
```
Comment (joshAppDev) [inline templates/setup/[stack]_system_monitor.js]: Can the stack name be read from the .env?

**Review by joshAppDev — COMMENTED** [review]
Looks good, and very comprehensive. I will need time to test & document.

**Comment by Hiro-Nakamura** [comment]
> Okay, I have fixed the test to utilize the new parameters. However, please note that the `--runtime` parameter must match the branch name of the 'ab-cli' branch. Otherwise, it will default to using `ab_runtime#master`.

Thanks for making those changes. 

Once all the various branches have been integrated, we don't need the `awsReady` branch anymore.  `#master` would be valid.

**Comment by Hiro-Nakamura** [comment]
> ### CLI vs Git
> The ideal setup in my mind is as simple as possible. Just git clone a repo, maybe copy & paste the .env settings, and then launch Docker. I really wasn't expecting ab-cli to be part of the AWS setup. Now I've never used AWS myself, but I imagine that you would type the .env settings into some Amazon interface, and it would propagate to all your EC2 instances, or something like that.
>

Yes, in the AWS implementation all the `.env` data will be entered in their `AWS Parameter Store`,  The scope of this particular set of changes however, was simply to remove the `config` container and pull all configuration data into environment variables.  With that change in place, it will be easy to move to their AWS implementation.

But we also need to make this available to our non AWS implementations as well.  And since we want to make this available for people outside of our team to easily install, the cli is a helpful tool to setup the environment for less technical people.


> 
> One benefit of deploying via a git repo, is that updating the site files (like docker-compose.yml) can be done by just a git pull.
> 

Good ideas.  

Technically speaking, that is what the cli is doing for you.  It is checking out the main data from our repo ( `ab_runtime` ) and then filling in the `.env` file with the data from your answers to the questions.  And auto `npm install`-ing for you, performing db migrations to the latest patches, and then updating the default Admin user data.

However, I didn't include the `docker-compose.*` files in the repo simply because I didn't feel comfortable leaving the database exposed if they said they didn't want that.  And handling that in the template was easier than performing a search and replace in an existing file.

But I like your idea about simply doing a git pull to get updates to the `docker-compose.*` files... So I could swap that around.

Along this same lines, would you also want a default `.env` file installed by the repo, then have the `cli` simply update/overwrite the `.env` with the data it collects from the questions?


> ### Database
> Will we be using an AWS DB for our database? If we are going to scale up to multiple instances, would they need to all connect to a central DB? (Again, I've never done this myself.)

Yes.  When we actually move this to AWS, we will remove our `db` container and replace the connection info to their DB server.  At the moment, I'm hoping our services will all be able to use the same connection settings rather than needing different logins each.  



### Non CLI new installs
So technically you could do a non cli install by doing the following:
- git clone the `ab_runtime` repo (once we include the docker compose files here)
- update the `.env` file with the changes
- perform a DB migration to bring the default install up to the latest changes.
- then perform an update to the default `admin` user so the new install doesn't have the default user/password/salt of the default install.


### Update existing install
Should be more like what you were describing if you are working with existing data:
- git clone the `ab_runtime` 
- update the `.env`

**Comment by Hiro-Nakamura** [comment]
@joshAppDev One thing I haven't done so far is to remove the file reference in `api_sails` for: 
```
    volumes:
      - type: bind
        source: ./logs/appbuilder/
        target: /var/log/appbuilder/
      - files:/data
```

I believe this was for the host system to monitor these logs and do some blacklisting of incoming connections.

I suppose we could simply remove this from our default configs, and only add it in if we have such a setup.  

Is there another way we could have this data shared without this file binding?

File: templates/setup/docker-compose.override.yml (hunk 1 above)
Comment (Hiro-Nakamura) [reply templates/setup/docker-compose.override.yml]: Yes, both port numbers are available in the `.env` file:
```
# External port numbers
WEB_PORT=<%= port %>
DB_PORT=<%= dbPort %>
```

We could simply update this to reference the ENV variable.

Do you know what the proper reference here would be for the ENV variable?

**Comment by joshAppDev** [comment]
> I believe this was for the host system to monitor these logs and do some blacklisting of incoming connections.
> I suppose we could simply remove this from our default configs, and only add it in if we have such a setup.

Yes, it should be fine to remove that log file binding. And I suppose AWS must have its own tools or services you can use to achieve the same level of security.

> Is there another way we could have this data shared without this file binding?

Maybe the log file could be exposed over an internal HTTP port, or written to a DB, which would then be exposed. But that seems to me like a bit too much overhead. We would need to develop both the part in the container that sends the log file, and the part outside the container that receives it. And then we would need to worry about locking it down so that it can't be read from outside of the server.

**Comment by joshAppDev** [comment]
I was thinking about how to control whether the DB is exposed. Since the Docker stack is launched through a shell script, maybe we can put the logic in there, with a .env setting for `EXPOSE_DB`. Something like:

```sh
if [ "$EXPOSE_DB" == "true" ]
then
    docker stack deploy -c docker-dompose.yml -c docker-compose.override.yml -c docker-compose.exposeDB.yml $STACKNAME
else
    docker stack deploy -c docker-dompose.yml -c docker-compose.override.yml $STACKNAME
fi
```
Then the person running the server can easily toggle the exposure on/off by changing the .env file.

**Comment by joshAppDev** [comment]
~~FYI, the installer is cloning the #master branch of ab-runtime, even though I added the option `--runtime=jh/awsReady` (assuming that's what this was for).~~

Turns out the `runtime` parameter name is case sensitive.

**Comment by joshAppDev** [comment]
Managed to fix the hang at "... watching log entries". The problem was the logs.js script failed to launch because of a Windows line ending compatibility issue with the `#!/usr/bin/env node` first line.

I changed my local copy of dockerStackWatch.js, to always proc this condition:
```js
               var command = path.join(process.cwd(), "logs.js");
               var options = [];
               if (true || os.platform() == "win32") {
                  options.push(command);
                  command = "node";
               }
```

**Comment by joshAppDev** [comment]
From what I can tell from my testing, it works. The site was installed, and I was able to log in. There were some errors in the console log about process manager, request timeouts, aborted DB connection while reading communication packets, etc. But I've often seen these types of errors on the production sites, so I'm guessing these are not because of the AWS changes.

# Changes when using podman

In general, the syntax fo podman and docker are nearly identical. So substitute all "docker" commands with "podman". There is no native `docker stack deploy -c *yml` equivalent in podman, but there is the `podman-compose` utility that helps to achieve the same thing. podman-compose is designed to emulate docker-compose, which is more or less like docker stack deploy.

There are a few more points to consider If we're talking about specifically about the NS server.
- We don't have root access on this server.
- We don't have ab-cli on this server.
- We can't run node.js outside a container.
- We have podman-compose installed, but are not using it directly. Doug set up systemd services for accounting2 and staging2. These services are what execute podman-compose for us. This setup allows for the containers to automatically start up when the server reboots. So that's why the start/stop scripts there are running `systemctl` instead of podman directly.

### Installing on the NS server

For accounting2 & staging2, the sites were installed by cloning the `ab-production-runtime` repo. The DB volumes were initialized manually. With the new changes towards .env config, future installs on the server could be done in a similar way with the `ab-runtime` repo. (As you described earlier.)

# Migrating a running site

Preface: I have not previously installed ABV2 sites myself using ab-cli, back before the AWS changes. So this is just  what I think should happen in theory.

Let's assume the running site was originally installed with `ab-cli`, on a server that runs Docker. That means the site files are based on the `ab-runtime` repo. So with a bit of work, we should be able to git pull the latest version of that repo.

### 1. Make sure ab-runtime#master is ready
The awsReady changes need to have been merged before sites can be migrated.

### 2. Backup config/local.js
It will be handy to reference this when configuring the .env file later.

Do NOT run `git pull` or `git fetch` at this point.

### 3. Take note of any other local modifications over ab-runtime.
```
git status
git diff
```
Backup anything important that has changed.

### 4. Revert changes to ab-runtime
For each file that you want to revert, run this:
```
git checkout -- <filename>
```
Delete the *.yml files, the UP/Down scripts. These will be replaced in the next step.

### 5. Pull
```
git pull
```
With all the changes already reverted, there should be no conflicts.

### 6. Configure
Edit the .env file.

And the good news is that future updates will be easy. Just git pull.

**Comment by joshAppDev** [comment]
Idea: Could we make user_manager auto create the admin account during the first time booting up, similar to how MariaDB does it? The initial admin email and password could be added to the .env file for this. In the base .sql init files, the admin password could be set to NULL or whatever, and user_manager looks for that, and replaces it.

**Review by joshAppDev — APPROVED** [review]

**Comment by Hiro-Nakamura** [comment]
You know, I like that idea about the `user_manager` performing an update on the site user when it boots up.  I'll look into adding that change.

**Pull request merged**